source/MaterialXGenGlsl/GlslResourceBindingContext.cpp

```diff
@@ -80,4 +80,84 @@ void GlslResourceBindingContext::emitResourceBindings(GenContext& context, const
     generator.emitLineBreak(stage);
 }
 
+void GlslResourceBindingContext::emitStructuredResourceBindings(GenContext& context,
+    const VariableBlock& uniforms, ShaderStage& stage, const std::string& structInstanceName,
+    const std::string& arraySuffix)
+{
+    ShaderGenerator& generator = context.getShaderGenerator();
+    const Syntax& syntax = generator.getSyntax();
+
+    // Glsl struct need to be aligned. We make a best effort to base align struct members and add
+    // padding if required
+    // https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_uniform_buffer_object.txt
+
+    const size_t baseAlignment = 16;
+    std::unordered_map<const TypeDesc*, size_t> alignmentMap({ { Type::FLOAT, baseAlignment / 4 },
+        { Type::INTEGER, baseAlignment / 4 }, { Type::BOOLEAN, baseAlignment / 4 },
+        { Type::COLOR2, baseAlignment }, { Type::COLOR3, baseAlignment },
+        { Type::COLOR4, baseAlignment }, { Type::VECTOR2, baseAlignment },
+        { Type::VECTOR3, baseAlignment }, { Type::VECTOR4, baseAlignment },
+        { Type::MATRIX33, baseAlignment * 4 }, { Type::MATRIX44, baseAlignment * 4 } });
+
+    //Get struct alignment and size
+    // alignment, uniform member index
+    vector<std::pair<size_t, size_t>> memberOrder;
+    size_t structSize = 0;
+    for (size_t i = 0; i < uniforms.size(); ++i)
+    {
+        auto it = alignmentMap.find(uniforms[i]->getType());
+        if (it == alignmentMap.end())
+        {
+            structSize += baseAlignment;
+            memberOrder.push_back(std::make_pair(baseAlignment, i));
+        }
+        else
+        {
+            structSize += it->second;
+            memberOrder.push_back(std::make_pair(it->second, i));
+        }
+    }
+
+    // Align up and determine number of padding floats to add
+    const size_t numPaddingfloats =
```

Comment:
This assumes that the uniform is 16-bit aligned if it's not of the types specified in the map.
We have float arrays so I think we should add that in even though without it it will be aligned but the uniform will not sorted properly however.

Reply:
For this current scope, I was assuming we won't have struct with arrays.

Reply:
```
struct structwithArray {
    vec3 a[20];
    vec2 b[10];
    float c;
};
```

Reply:
As discussed this is more complicated that assumed. We can park this for later.